### PR TITLE
Add math & bitwise ops to U256 type 2

### PR DESCRIPTION
**Pull request by rostyslavtyshko**
No new changes compared [to existing PR about math & bitwise ops to U256](https://github.com/FuelLabs/sway/pull/2283) created only due to rebaseing issues and already solved compiler issue

### CODE REVIEW

**Comment by JoshuaBatty** [comment]
Hi @rostyslavtyshko I've spent today looking into why this test isn't passing. It's actually happening in the sway compiler and crashing with a stack over flow message. 

With the `u256` lib in the std-lib we are *just* sneaking in under the maximum stack size when compiling, but when the compiler is invoked from a test (such as in LSP tests) the stack is getting tipped over when some of the stack is also being used for the test. 

I think perhaps the parser needs to be optimized slightly before we can land this PR unfortunately.

**Review by adlerjohn — COMMENTED** [review]
Is this PR meant to supersede #2283?

File: sway-lib-core/src/ops.sw [hunk 1]
```diff
@@ -618,15 +618,15 @@ impl Shiftable for b256 {
 /////////////////////////////////////////////////
 
 /// Left shift a u64 and preserve the overflow amount if any
-fn lsh_with_carry(word: u64, shift_amount: u64) -> (u64, u64) {
+pub fn lsh_with_carry(word: u64, shift_amount: u64) -> (u64, u64) {
```
Comment (adlerjohn) [inline sway-lib-core/src/ops.sw:621]: `core` is meant to just contain ops, necessary so `std` can use operators. Everything else in `core` should be private. If you need access to these two helper functions, then copy them into `std`.

The only other thing so far that's acceptable is `core::num`, to imitate Rust's https://doc.rust-lang.org/src/core/num/mod.rs.html#849-853 (and even then we should consider moving that to `std` as per https://doc.rust-lang.org/std/primitive.u64.html#associatedconstant.MAX).

There was a suggestion to add a second library in between `core` and `std`, but I can't find it. cc @nfurfaro

File: sway-lib-std/src/u256.sw [hunk 2]
```diff
@@ -89,4 +91,512 @@ impl U256 {
...
+            i -= 1;
+        }
+
+        quotient
+    }
+}
+
+impl core::ops::Ord for U256 {
+    fn gt(self, other: Self) -> bool {
+        self.a > other.a || self.a == other.a && self.b > other.b ||  self.a == other.a && self.b == other.b  && self.c > other.c ||  self.a == other.a && self.b == other.b  && self.c == other.c && self.d > other.d
```
Comment (adlerjohn) [inline sway-lib-std/src/u256.sw:356]: Can you put brackets on this one so it's formatted the same as `lt` and `eq`?

File: sway-lib-std/src/u256.sw [hunk 3]
```diff
@@ -89,4 +91,512 @@ impl U256 {
...
+            w3 = shifted_1;
+        } else if w == 3 {
+            w4 = word_1 >> b;
+        };
+
+        ~U256::from(w1, w2, w3, w4)
+    }
+}
+
+impl core::ops::Add for U256 {
```
Comment (adlerjohn) [inline sway-lib-std/src/u256.sw:475]: Wait...why is there a duplicate here?

**Review by Braqzen — COMMENTED** [review]
Spacing should be consistent. 
If the formatter doesn't spot this then could you raise an issue so that `let(a, b, c)` gets formatted to `let (a, b, c)`?

**Review by adlerjohn — APPROVED** [review]

**Review by Braqzen — COMMENTED** [review]
I think the tests should also test `a` & `b` to prove that their values are correct.

**Review by sezna — APPROVED** [review]

**Pull request merged**